types/i18next/index.d.ts

```diff
@@ -1,157 +1,601 @@
+        /**
+         * object with vars for interpolation - or put them directly in options
+         */
         replace?: any;
+        /**
+         * override language to use
+         */
         lng?: string;
+        /**
+         * override languages to use
+         */
         lngs?: string[];
+        /**
+         * override language to lookup key if not found see fallbacks for details
+         */
         fallbackLng?: FallbackLng;
+        /**
+         * override namespaces (string or array)
+         */
         ns?: string | string[];
+        /**
+         * override char to separate keys
+         */
         keySeparator?: string;
+        /**
+         * override char to split namespace from key
+         */
         nsSeparator?: string;
+        /**
+         * accessing an object not a translation string (can be set globally too)
+         */
         returnObjects?: boolean;
+        /**
+         * char, eg. '\n' that arrays will be joined by (can be set globally too)
+         */
         joinArrays?: string;
-        postProcess?: string | any[];
+        /**
+         * string or array of postProcessors to apply see interval plurals as a sample
+         */
+        postProcess?: string | string[];
+        /**
+         * override interpolation options
+         */
         interpolation?: InterpolationOptions;
-        //add an indexer to assure that interpolation arguments can be passed
-        [x: string]: any;
     }
 
-    interface Options {
-        debug?: boolean;
-        resources?: ResourceStore;
-        lng?: string;
-        fallbackLng?: FallbackLng;
-        ns?: string | string[];
-        defaultNS?: string;
-        fallbackNS?: string | string[];
-        whitelist?: string[];
-        lowerCaseLng?: boolean;
-        load?: string
-        preload?: string[];
-        keySeparator?: string | false;
-        nsSeparator?: string | false;
-        pluralSeparator?: string;
-        contextSeparator?: string;
-        saveMissing?: boolean;
-        saveMissingTo?: string;
-        missingKeyHandler?: (lng: string, ns: string, key: string, fallbackValue: string) => void;
-        parseMissingKeyHandler?: (key: string) => void;
-        appendNamespaceToMissingKey?: boolean;
-        postProcess?: string | any[];
-        returnNull?: boolean;
-        returnEmptyString?: boolean;
-        returnObjects?: boolean;
-        returnedObjectHandler?: (key: string, value: string, options: any) => void;
-        joinArrays?: string;
-        overloadTranslationOptionHandler?: (args: any[]) => TranslationOptions;
-        interpolation?: InterpolationOptions;
-        detection?: any;
-        backend?: any;
-        cache?: any;
+    type Callback = (error: any, t: TranslationFunction) => void;
+
+    type TranslationFunction<TResult = string, TValues extends object = object, TKeys extends string = string> = (key: TKeys | TKeys[], options?: TranslationOptions<TValues>) => TResult;
```

Comment:
It's a nice idea to restrict `t()` function to certain keys and options. `TResult` should be `any` by default (we discussed that in previous review).